contrib/parseq-restli-client/src/main/java/com/linkedin/restli/client/ParSeqRestClient.java

```diff
@@ -197,10 +207,49 @@ static String generateTaskName(final Request<?> request) {
   private <T> Task<Response<T>> createTask(final String name, final Request<T> request,
       final RequestContext requestContext, RequestConfig config) {
     LOGGER.debug("createTask, name: '{}', config: {}", name, config);
+    return createTaskWithTimeout(name, request, requestContext, config);
+  }
+
+  // reconcile timeout specified in requestContext with the given timeout
+  private void reconcileRequestTimeout(final RequestContext requestContext, Long timeout) {
+    Number contextTimeout = (Number) requestContext.getLocalAttr(R2Constants.REQUEST_TIMEOUT);
+    int timeoutBound = (contextTimeout != null) ? contextTimeout.intValue() : Integer.MAX_VALUE;
+    // d2 request timeout can only accept Integer timeout
+    requestContext.putLocalAttr(R2Constants.REQUEST_TIMEOUT, Math.min(timeout, timeoutBound));
+  }
+
+  private <T> Task<Response<T>> createTaskWithTimeout(final String name, final Request<T> request,
+      final RequestContext requestContext, RequestConfig config) {
+    ConfigValue<Long> timeout = config.getTimeoutMs();
+    if (timeout.getValue() != null && timeout.getValue() > 0 && ParSeqGlobalConfiguration.isD2RequestTimeoutEnabled()) {
+      // reconcile timeout if request has already set timeout in its request context
+      reconcileRequestTimeout(requestContext, timeout.getValue());
+    }
+    Task<Response<T>> requestTask;
     if (RequestGroup.isBatchable(request, config)) {
-      return withTimeout(createBatchableTask(name, request, requestContext, config), config);
+      requestTask = createBatchableTask(name, request, requestContext, config);
+    } else {
+      requestTask = Task.async(name, () -> sendRequest(request, requestContext));
+    }
+    if (!ParSeqGlobalConfiguration.isD2RequestTimeoutEnabled()) {
+      return withTimeout(requestTask, timeout);
     } else {
-      return withTimeout(Task.async(name, () -> sendRequest(request, requestContext)), config);
+      // transform TimeoutException to be compatible with current Task timeout exception message
+      return requestTask.transform((Try<Response<T>> tryGet) -> {
```

Comment:
We should get this transform Task a name - preferably the same as the one that original withTimeout had i.e. so that it includes the configuration key that caused timeout to be applied.